src/mumble/pluginAPI/Plugin.h

```diff
@@ -0,0 +1,139 @@
+	///
+	/// @returns A pointer to the description (encoded as a C-String)
+	PLUGIN_EXPORT const char* getDescription();
+
+	/// Provides the MumbleAPI struct to the plugin. This struct contains function pointers that can be used
+	/// to interact with the Mumble client. It is up to the plugin to store this struct somewhere if it wants to make use
+	/// of it at some point.
+	///
+	/// @param api The MumbleAPI struct
+	PLUGIN_EXPORT void registerAPIFunctions(const struct MumbleAPI *api);
+
+	/// Registers the ID of this plugin. This is the ID Mumble will reference this plugin with and by which this plugin
+	/// can identify itself when communicating with Mumble.
+	///
+	/// @param id The ID for this plugin
+	PLUGIN_EXPORT void registerPluginID(uint32_t id);
+
+
+	// Parameters to functions below are yet to be determined
+
+	// -------- Callback functions -----------
+
+	/// Called when connecting to a server.
+	PLUGIN_EXPORT void onServerConnected();
+
+	/// Called when disconnecting from a server.
+	PLUGIN_EXPORT void onServerDisconnected();
+
+	/// Called whenever any user on the server enters a channel
+	/// This function will also be called when freshly connecting to a server as each user on that
+	/// server needs to be "added" to the respective channel as far as the local client is concerned.
+	///
+	/// @param userID The ID of the user this event has been triggered for
+	/// @param previousChannelID The ID of the chanel the user is coming from. Negative IDs indicate that there is no previous channel (e.g. the user
+	/// 	freshly connected to the server) or the channel isn't available because of any other reason.
+	/// @param newChannelID The ID of the channel the user has entered. If the ID is negative, the new channel could not be retrieved. This means
+	/// 	that the ID is invalid.
+	PLUGIN_EXPORT void onChannelEntered(MumbleUserID_t userID, MumbleChannelID_t previousChannelID, MumbleChannelID_t newChannelID);
+
+	/// Called whenever a user leaves a channel.
+	/// This function will also be called when the user disconnects from the server.
+	///
+	/// @param userID The ID of the user that left the channel
+	/// @param channelID The ID of the channel the user left. If the ID is negative, the channel could not be retrieved. This means that the ID is
+	/// 	invalid.
+	PLUGIN_EXPORT void onChannelExited(MumbleUserID_t userID, MumbleChannelID_t channelID);
+
+	/// Called when any user changes his/her talking state.
+	///
+	/// @param userID The ID of the user whose talking state has been changed
+	/// @param talkingState The new TalkingState the user has switched to.
+	PLUGIN_EXPORT void onUserTalkingStateChanged(MumbleUserID_t userID, TalkingState_t talkingState);
+
+	/// Called whenever there is audio input
+	PLUGIN_EXPORT void onAudioInput(short *inputPCM, bool isSpeech);
+
+	/// Called whenever there is audio output
+	PLUGIN_EXPORT void onAudioOutput_short(short *outputPCM, uint32_t sampleCount, uint32_t channelCount);
```

Comment:
I think uint8_t is enough for channel count.
Also not sure currently but we might want a return value indicating whether the stream was modified.
If mumble needs to convert from float internal to short representation just for us. It would be a huge waste of performance to convert all the way back even though nothing was changed.

Reply:
> I think uint8_t is enough for channel count.

Internally a plain `int` is used, though I guess we will never encounter a server that has so many channels that we'll actually need a full int type for it.

On the other hand 256 channels (which is what could be represented with `uint8_t` right?) might seem a lot but not yet unreachable. I think that I wouldn't want to go below `uint16_t` for this one...

> If mumble needs to convert from float internal to short representation just for us. It would be a huge waste of performance to convert all the way back even though nothing was changed.

The way I intend to implement this callback won't involve back-converting. Depending on some internal state (which I believe is ultimately dependent on what kind of audio backend is being used), Mumble either processes the audio stream as a float- or a short-array. So it needs that specific representation anyways. Thus only one of the two callbacks will be used for a particular backend anyways.
That way no back-conversion should be necessary.